docs/writing-stories/typescript.md

```diff
@@ -0,0 +1,149 @@
+---
+title: 'Writing stories in TypeScript'
+---
+
+Writing your stories in [TypeScript](https://www.typescriptlang.org/) makes you more productive. You don't have to jump between files to look up component props. Your code editor will alert you about missing required props and even autocomplete prop values. Plus, Storybook infers those component types to auto-generate the [Controls](../api/doc-block-controls.md) table.
+
+Using TypeScript also makes your code more robust. When authoring stories, you're replicating how a component will be used within your app. With type checking, you can catch bugs and edge cases as you code.
+
+Storybook has built-in TypeScript support, so you can get started with zero configuration required.
+
+## Typing stories with `Meta` and `StoryObj`
+
+When writing stories, there are two aspects that are helpful to type. The first is the [component meta](./introduction.md#default-export), which describes and configures the component and its stories. In a [CSF file](..api.md/csf), this is the default export. The second is the [stories themselves](./introduction.md#defining-stories).
+
+Storybook provides utility types for each of these, named `Meta` and `StoryObj`. Here's an example CSF file using those types:
+
+<!-- prettier-ignore-start -->
+
+<CodeSnippets
+  paths={[
+    'angular/typed-csf-file.ts.mdx',
+    'web-components/typed-csf-file.ts.mdx',
+    'common/typed-csf-file.ts.mdx',
+  ]}
+/>
+
+<!-- prettier-ignore-end -->
+
+### Optional type parameter
+
+`Meta` and `StoryObj` types are both [generics](https://www.typescriptlang.org/docs/handbook/2/generics.html#working-with-generic-type-variables), so you can provide them with a prop type parameter for the component type or the component's props type (e.g., the `typeof Button` portion of `Meta<typeof Button>`). By doing so, TypeScript will prevent you from defining an invalid arg, and all [decorators](./decorators.md), [play functions](./play-function.md), or [loaders](./loaders.md) will type their function arguments.
+
+The example above passes a component type. See [**Typing custom args**](#typing-custom-args) for an example of passing a props type.
+
+## Using `satisfies` for better type safety
+
+<IfRenderer renderer={['angular', 'web-components']}>
+
+<div class="aside">
+
+ℹ️ We are not yet able to provide additional type safety using the `satisfies` operator with Angular and Web components.
+
+<details>
+<summary>More info</summary>
+
+Both Angular and Web components utilize a class plus decorator approach. The decorators provide runtime metadata, but do not offer metadata at compile time.
+
+As a result, it appears impossible to determine if a property in the class is a required property or an optional property (but non-nullable due to a default value) or a non-nullable internal state variable.
+
+For more information, please refer to [this discussion](github.com/storybookjs/storybook/discussions/20988).
+
+</details>
```

Comment:
@kylegach while previewing this locally, it reads a bit weird and adds a bit of confusion with the amount of information you're trying to present. A couple of options we could here:
1- Condense the information in a single paragraph without the aside and provide the link for the relevant discussion (don't forget to update the link as it throws a 404).
2 - Push this aside into a troubleshooting section visible only for the frameworks (i.e., Angular and Web Components) with some word adjustments

Reply:
I fixed the link (thanks for the catch!). But I still like the current approach better than either of those options.

 - The aside makes it immediately clear to the affected users that `satisfies` is not available.
 - Rather than just taking our word for it, the "More info" provides the _why_.
     - While this could be an anchor link to a conditionally rendered troubleshooting section, that seemed far more clunky than a details disclosure.
 - I don't think we should combine the initial message with what's in "More info", because _that_ is too much information to present upfront.